src/Workspaces/Core/Portable/Utilities/CancellationSeries.cs

```diff
@@ -0,0 +1,130 @@
+    /// Produces a series of <see cref="CancellationToken"/> objects such that requesting a new token
+    /// causes the previously issued token to be cancelled.
+    /// </summary>
+    /// <remarks>
+    /// <para>Consuming code is responsible for managing overlapping asynchronous operations.</para>
+    /// <para>This class has a lock-free implementation to minimise latency and contention.</para>
+    /// </remarks>
+    internal sealed class CancellationSeries : IDisposable
+    {
+        private CancellationTokenSource? _cts = new();
+
+        private readonly CancellationToken _superToken;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CancellationSeries"/>.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the last
+        /// issued token and causes any subsequent tokens to be issued in a cancelled state.</param>
+        public CancellationSeries(CancellationToken token = default)
+        {
+            _superToken = token;
+
+#if DEBUG
+            _ctorStack = new StackTrace();
+#endif
+        }
+
+#if DEBUG
+        private readonly StackTrace _ctorStack;
+
+        ~CancellationSeries()
+        {
+            Contract.ThrowIfFalse(
+                Environment.HasShutdownStarted || _cts == null,
+                $"Instance of CancellationSeries not disposed before being finalized{Environment.NewLine}Stack at construction:{Environment.NewLine}{_ctorStack}");
+        }
+#endif
+
+        /// <summary>
+        /// Creates the next <see cref="CancellationToken"/> in the series, ensuring the last issued
+        /// token (if any) is cancelled first.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the
+        /// returned token.</param>
+        /// <returns>
+        /// A cancellation token that will be cancelled when either:
+        /// <list type="bullet">
+        /// <item><see cref="CreateNext"/> is called again</item>
+        /// <item>The token passed to this method (if any) is cancelled</item>
+        /// <item>The token passed to the constructor (if any) is cancelled</item>
+        /// <item><see cref="Dispose"/> is called</item>
+        /// </list>
+        /// </returns>
+        /// <exception cref="ObjectDisposedException">This object has been disposed.</exception>
+        public CancellationToken CreateNext(CancellationToken token = default)
+        {
+            var nextSource = CancellationTokenSource.CreateLinkedTokenSource(token, _superToken);
+
+            // Obtain the token before exchange, as otherwise the CTS may be cancelled before
+            // we request the Token, which will result in an ObjectDisposedException.
+            // This way we would return a cancelled token, which is reasonable.
+            var nextToken = nextSource.Token;
+
+            var priorSource = Interlocked.Exchange(ref _cts, nextSource);
```

Comment:
i noticed a loop in this code for another version taht i was looking at.  is there no need for that?  did that go away?  are we deviating from the code for a good reason?  or would it just be good to keep both entirely in sync?

Reply:
This code was pulled from dotnet/project-system, and fails to avoid state corruption when `CreateNext` is called after the object is disposed.

Reply:
➡️ I restored the latch